src/python/pants/source/source_root.py

```diff
@@ -266,47 +297,43 @@ def register_options(cls, register):
             "`<buildroot>/project1/src/python`.  A `*` wildcard will match a single path segment, "
             "e.g., `src/*` will match `<buildroot>/src/python` and `<buildroot>/src/rust`. "
-            "Use `^` to signify that the buildroot itself is a source root. "
+            "Use `/` to signify that the buildroot itself is a source root. "
             "See https://pants.readme.io/docs/source-roots.",
         )
 
     @memoized_method
     def get_source_roots(self):
-        return SourceRoots(self)
+        return SourceRoots(self.options.root_patterns, self.options.unmatched == "fail", self)
 
     def create_trie(self) -> "SourceRootTrie":
         """Create a trie of source root patterns from options."""
         trie = SourceRootTrie()
         options = self.get_options()
 
-        if options.root_patterns:
-            for path in options.root_patterns:
-                trie.add_pattern(path)
-        else:
-            legacy_patterns = []
-            for category in ["source", "test", "thirdparty"]:
-                # Add patterns.
-                for pattern in options.get("{}_root_patterns".format(category), []):
-                    trie.add_pattern(pattern)
-                    legacy_patterns.append(pattern)
-                # Add fixed source roots.
-                for path, langs in options.get("{}_roots".format(category), {}).items():
-                    trie.add_fixed(path)
-                    legacy_patterns.append(f"^/{path}")
-            # We need to issue a deprecation warning even if relying on the default values
-            # of the deprecated options.
-            deprecated_conditional(
-                lambda: True,
-                removal_version="1.30.0.dev0",
-                entity_description="the *_root_patterns and *_roots options",
-                hint_message="Explicitly list your source roots with the `root_patterns` option in "
-                "the [source] scope. See https://pants.readme.io/docs/source-roots. "
-                f"Your current roots are covered by: [{', '.join(legacy_patterns)}]",
-            )
+        legacy_patterns = []
+        for category in ["source", "test", "thirdparty"]:
+            # Add patterns.
+            for pattern in options.get("{}_root_patterns".format(category), []):
+                trie.add_pattern(pattern)
+                legacy_patterns.append(pattern)
+            # Add fixed source roots.
+            for path, langs in options.get("{}_roots".format(category), {}).items():
+                trie.add_fixed(path)
+                legacy_patterns.append(f"^/{path}")
+        # We need to issue a deprecation warning even if relying on the default values
+        # of the deprecated options.
+        deprecated_conditional(
+            lambda: True,
+            removal_version="1.30.0.dev0",
+            entity_description="the *_root_patterns and *_roots options",
```

Comment:
Stale?

Reply:
Those are still deprecated - those are the old {source,test,3rdparty}- prefixed options.